Implement new popover inside invoker behaviour

lukewarlow opened the pull request:
#### 8a02dacf8e5b11135f826944889ac535264f3e41
<pre>
Implement new popover inside invoker behaviour
<a href="https://bugs.webkit.org/show_bug.cgi?id=283494">https://bugs.webkit.org/show_bug.cgi?id=283494</a>

Reviewed by Tim Nguyen.

This patch matches the new spec behaviour for when a popover is inside its invoker.

See <a href="https://github.com/whatwg/html/pull/10770">https://github.com/whatwg/html/pull/10770</a>

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-nested-in-button-expected.txt:
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::defaultEventHandler):
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::handlePopoverTargetAction const):
* Source/WebCore/html/HTMLFormControlElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::defaultEventHandler):

Canonical link: <a href="https://commits.webkit.org/287522@main">https://commits.webkit.org/287522@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/22aa5923ca43ff37522ebb48eae74a8a8cadc459

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79928 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58925 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33325 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84442 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30902 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82036 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67989 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7220 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62464 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20291 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82995 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52530 "Found 1 new test failure: media/audioSession/getUserMedia.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72789 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42775 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49866 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26938 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29364 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70986 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27418 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85874 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7144 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5010 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70738 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7321 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68629 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69981 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17443 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13979 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12911 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7108 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12643 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6955 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10467 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8759 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -395,14 +395,19 @@ void HTMLFormControlElement::setPopoverTargetAction(const AtomString& value)
-void HTMLFormControlElement::handlePopoverTargetAction() const
+void HTMLFormControlElement::handlePopoverTargetAction(const EventTarget* eventTarget) const
 {
     RefPtr target = popoverTargetElement();
```
Can you rename this variable to `popover` to make the logic less confusing to read?

nt1m commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -395,14 +395,19 @@ void HTMLFormControlElement::setPopoverTargetAction(const AtomString& value)
 
     ASSERT(target->popoverData());
 
+    if (RefPtr eventTargetElement = dynamicDowncast<Node>(eventTarget)) {
```
I'd name this `eventTargetNode`

nt1m commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -395,14 +395,19 @@ void HTMLFormControlElement::setPopoverTargetAction(const AtomString& value)
     ASSERT(target->popoverData());
 
+    if (RefPtr eventTargetElement = dynamicDowncast<Node>(eventTarget)) {
+        if (target->containsIncludingShadowDOM(eventTargetElement.get()) && target.get()->isDescendantOrShadowDescendantOf(this))
```
```suggestion
        if (target->containsIncludingShadowDOM(eventTargetElement) && target->isDescendantOrShadowDescendantOf(this))
```

You shouldn't need to use `.get()` here.

lukewarlow replied:
It seems I do need eventTargetElement.get() (see below), but I don't need ~~target~~popover.get() so I'll remove that.

```
/Source/WebCore/html/HTMLFormControlElement.cpp:407:49: error: no viable conversion from
      'RefPtr<const Node>' (aka 'WTF::RefPtr<const WebCore::Node>') to 'const Node *'
  407 |         if (popover->containsIncludingShadowDOM(eventTargetNode) && popover->isDescendantOrShadowDescendantOf(this))
      |                                                 ^~~~~~~~~~~~~~~
...
/Users/luke/Development/BrowserEngines/WebKit/Source/WebCore/dom/Node.h:444:63: note: passing argument to parameter here
  444 |     WEBCORE_EXPORT bool containsIncludingShadowDOM(const Node*) const;
  ```

lukewarlow committed:
Implement new popover inside invoker behaviour
https://bugs.webkit.org/show_bug.cgi?id=283494

Reviewed by Tim Nguyen.

This patch matches the new spec behaviour for when a popover is inside its invoker.

See https://github.com/whatwg/html/pull/10770

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-nested-in-button-expected.txt:
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::defaultEventHandler):
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::handlePopoverTargetAction const):
* Source/WebCore/html/HTMLFormControlElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::defaultEventHandler):

Canonical link: https://commits.webkit.org/287522@main
